src/server/ua_services_attribute.c

```diff
@@ -1097,16 +1097,29 @@ writeValueAttribute(UA_Server *server, UA_Session *session,
         else
             retval = writeValueAttributeWithRange(node, &adjustedValue, rangeptr);
 
+#ifdef UA_ENABLE_HISTORIZING
+        if(retval == UA_STATUSCODE_GOOD && server->config.historyDataService.setValue)
+            server->config.historyDataService.setValue(server->config.historyDataService.context,
+                                                       server,
+                                                       session ? &session->sessionId : NULL,
+                                                       session ? session->sessionHandle : NULL,
+                                                       &node->nodeId,
+                                                       &adjustedValue);
+#endif
         /* Callback after writing */
         if(retval == UA_STATUSCODE_GOOD && node->value.data.callback.onWrite)
-            node->value.data.callback.onWrite(server, &session->sessionId,
-                                              session->sessionHandle, &node->nodeId,
+            node->value.data.callback.onWrite(server,
+                                              session ? &session->sessionId : NULL,
```

Comment:
There is always a session. We the plugin should not have to check for NULL here.
This can be state in the API docs.

We could add an assertion that session != NULL. to the service.